### ISSUE TITLE
print traceback in successResultOf error reporting

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6377]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#6377|
|Type|enhancement|
|Created|2013-03-18 19:21:59Z|
|Branch|https://github.com/twisted/twisted/tree/successResultOf-traceback-6377|

It currently prints the `repr` of the `Failure`, it'd be nice to be able to find out where the `Failure` occurred.

Also, would be nice to be able to pass optional errors to `failureResultOf` to check which errors are supposed to have been raised similar to (like `assertRaises`)

<details><summary>Searchable metadata</summary>

```
trac-id__6377 6377
type__enhancement enhancement
reporter__cyli cyli
priority__normal normal
milestone__ 
branch__branches_successResultOf_traceback_6377 branches/successResultOf-traceback-6377
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1363634519000000 1363634519000000
changetime__1364366706000000 1364366706000000
version__None None
owner__cyli cyli

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli commented|
|-|-|

(In [37591]) Branching to 'successResultOf-traceback-6377'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

(In [37598]) Add an optional expected exception type to failureResultOf.  refs [#6377](https://github.com/twisted/twisted/issues/6377)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

Forgot to include a refs tag in the commit message in the previous commit:

(In r37595) Include traceback in successResultOf error reporting

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [37599]) Add topfile.  refs [#6377](https://github.com/twisted/twisted/issues/6377)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli removed owner|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/successResultOf-traceback-6377)

The failure thing might be out of scope and perhaps should be broken into a different ticket, but these are pretty tiny changes.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince set owner to @cyli|
|-|-|

 - `successResultOf` and `failureResultOf` are supported on python 3, so additional behaviour for them should be tested, and tested succesfully.
 - I'd be inclined so say that `failureResultOf` should take a list of exception types, like `Failure.trap`, `Failure.check`, `TestCase.assertFailure`. (This would incidentally allow you to simply pass the list to `Failure.check`, rather than 'if x and f.check(x):`.
 - The news item would read better as two seperate news items, which suggests that this should be two tickets.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

(In [37616]) Move failureResultOf changes to a different ticket, and fix python3 tests.  refs [#6377](https://github.com/twisted/twisted/issues/6377)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli removed owner|
|-|-|

Thanks for the quick review tom.prince!

 1.  Hm... ok, I'm now testing to see if the traceback is in the str of the exception, since in Python 3 exceptions don't necessarily have a `.message` attribute anymore.
 1.  I've split the `failureResultOf` stuff into [#6380](https://github.com/twisted/twisted/issues/6380).
 1.  Changed the newsfile to a `.misc` instead.

[buildbot results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/successResultOf-traceback-6377/)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince set owner to @cyli|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli set status to `closed`|
|-|-|

(In [37828]) Merge successResultOf-traceback-6377: more useful traceback if failure

Author: cyli
Reviewer: tom.prince
Fixes: [#6377](https://github.com/twisted/twisted/issues/6377)

If the Deferred provided to successResultOf errbacks, print a traceback instead of just the repr of the failure.
